src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs

```diff
@@ -397,27 +367,355 @@ CollectionExpressionTypeKind.Span or
+                rewrittenReceiver = _factory.New(constructor, ImmutableArray.Create(GetKnownLengthExpression(elements, numberIncludingLastSpread, localsBuilder)));
+            }
+            else
+            {
+                // List<ElementType> list = new();
+                var constructor = ((MethodSymbol)_factory.WellKnownMember(WellKnownMember.System_Collections_Generic_List_T__ctor)).AsMember(collectionType);
+                rewrittenReceiver = _factory.New(constructor, ImmutableArray<BoundExpression>.Empty);
+            }
+
+            // Create a temp for the list.
+            BoundAssignmentOperator assignmentToTemp;
+            BoundLocal listTemp = _factory.StoreToTemp(rewrittenReceiver, out assignmentToTemp, isKnownToReferToTempIfReferenceType: true);
+            localsBuilder.Add(listTemp);
+            sideEffects.Add(assignmentToTemp);
+
+            var addMethod = ((MethodSymbol)_factory.WellKnownMember(WellKnownMember.System_Collections_Generic_List_T__Add)).AsMember(collectionType);
+            AddCollectionExpressionElements(
+                elements,
+                listTemp,
+                localsBuilder,
+                numberIncludingLastSpread,
+                sideEffects,
+                (ArrayBuilder<BoundExpression> expressions, BoundExpression listTemp, BoundExpression rewrittenValue) =>
+                {
+                    // list.Add(element);
+                    expressions.Add(
+                        _factory.Call(listTemp, addMethod, rewrittenValue));
+                });
+
+            var locals = localsBuilder.SelectAsArray(l => l.LocalSymbol);
+            localsBuilder.Free();
+
+            return new BoundSequence(
+                node.Syntax,
+                locals,
+                sideEffects.ToImmutableAndFree(),
+                listTemp,
+                collectionType);
+        }
+
+        private BoundExpression RewriteCollectionExpressionElementExpression(BoundExpression element)
+        {
+            var expression = element is BoundCollectionExpressionSpreadElement spreadElement ?
+                spreadElement.Expression :
+                element;
+            return VisitExpression(expression);
+        }
+
+        private void RewriteCollectionExpressionElementsIntoTemporaries(
+            ImmutableArray<BoundExpression> elements,
+            int numberIncludingLastSpread,
+            ArrayBuilder<BoundLocal> locals,
+            ArrayBuilder<BoundExpression> sideEffects)
+        {
+            for (int i = 0; i < numberIncludingLastSpread; i++)
+            {
+                var rewrittenExpression = RewriteCollectionExpressionElementExpression(elements[i]);
+                BoundAssignmentOperator assignmentToTemp;
+                BoundLocal temp = _factory.StoreToTemp(rewrittenExpression, out assignmentToTemp, isKnownToReferToTempIfReferenceType: true);
```

Comment:
Ideally, we shouldn't require a temporary for an element that has a constant value.